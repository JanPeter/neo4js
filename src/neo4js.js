import { v1 as neo4j } from 'neo4j-driver';
import Utils from './Utils';
import uuid from 'node-uuid';
import Debug from 'debug';
const debug = Debug('neo4js');

import * as neo4jsErrors from './errors';
import Model from './model';
import ModelManager from './model-manager';
import ModelObject from './model-object';
import Query from './query';
import { getSchemaInfo } from './schema-info';

class Neo4js {
  /**
   * Constructor to create a new instance of the Neo4js class
   * @param {String}  username The username which is used to authenticate against the database
   * @param {String}  password The password which is used to authenticate against the database
   * @param {Object}  [options={}] An object with options
   * @param {String}  [options.url=localhost] Bolt url
   * @param {Integer} [options.port=7687] Bolt port
   * @param {Object}  [options.rest={}]
   * @param {String}  [options.rest.url=localhost]
   * @param {Integer} [options.rest.port=7474]
   * @param {Integer} [options.connectionPoolSize=50]
   */
  constructor(username, password, options) {
    this.auth = { username, password };

    const defaultOptions = {
      url: `bolt://localhost`,
      port: 7687,
      rest: {
        url: `http://localhost`,
        port: 7474,
      },
      driver: {
        connectionPoolSize: 50,
      },
    };

    options = Utils._.extend({}, defaultOptions, options || {});
    options.rest = Utils._.extend({}, defaultOptions.rest, options.rest);
    options.driver = Utils._.extend({}, defaultOptions.driver, options.driver);

    // TODO: Maybe add here some more checks...
    const url = `${options.url}:${options.port}`;
    this.driver = neo4j.driver(url, neo4j.auth.basic(username, password), options.driver);

    this.options = options;

    this.models = {};
    this.modelManager = new ModelManager(this);
  }

  /**
   * @param {String | String[]} labels
   * @param {Object} schema
   * @param {Object} schema.{propertyName}
   * @param {Boolean} schema.{propertyName}.index
   * @param {Boolean} schema.{propertyName}.unique
   * @param {Boolean} schema.{propertyName}.exists
   * @param {Function} schema.{propertyName}.defaultValue
   */
  define(labels, schema, hooks) {
    const model = new Model(labels, schema, hooks, this);
    this.modelManager.add(model);
    return model;
  }

  /**
   * @param {String | String[]} labels
   */
  getModel(labels) {
    const model = this.modelManager.getModel(labels);
    return model;
  }

  /**
   * @returns {Promise}
   */
  getSchemaInfo() {
    return getSchemaInfo(this.options.rest.url, this.options.rest.port, this.auth);
  }

  drop() {
    return this.getSchemaInfo()
      .then(result => {
        const cmds = [];
        for (const label in result) {
          for (const p in result[label]) {
            if (result[label][p].index) {
              cmds.push(new Query(this).dropIndex(label, p));
            }
            if (result[label][p].unique) {
              cmds.push(new Query(this).dropConstraint(label, p, 'unique'));
            }
          }
        }
        cmds.push(new Query(this).match('n').detach('n'));
        return Query.runSequence(cmds);
      });
  }

  /**
   * @param {Object} options
   * @param {Boolean} options.force
   */
  sync(options) {
    debug(`sync start`);
    options = options || {};

    return Promise.resolve()
      .then(() => {
        if (options.force) {
          debug(`cleanup database start`);
          return this.drop();
        }
      })
      .then(() => {
        if (options.force) {
          debug(`cleanup database end`);
        }
        const promises = this.modelManager.all.map(m => m.sync());
        return Promise.all(promises);
      })
      .then(syncResults => {
        const errors = [];
        for (const results of syncResults) {
          for (const result of results) {
            if (result.err) {
              errors.push(result.err);
            }
          }
        }
        return errors;
      })
      .then(result => {
        debug(`sync end`);
        return result;
      })
      .catch(err => {
        debug(`sync end with errors`, err);
        return err;
      });
  }

  /**
   * This method creates a new instance of a Query
   * @return {Query}
   */
  Query() {
    return new Query(this);
  }

  /**
   * This method runs a command with params
   * @param {String} cmd - Cypher statement to execute
   * @param {Object} params - Map with parameters to use in statement
   * @return {Promise} Resolves either with the result or the error message
   */
  run(cmd, params) {
    const session = this.driver.session();
    return session.run(cmd, params)
      .then(result => {
        session.close();
        return result;
      })
      .catch(err => {
        session.close();
        return err;
      });
  }

  /**
   * @returns {UUIDv4}
   */
  static uuid4() {
    return uuid.v4();
  }

  /**
   * @returns {UUIDv1}
   */
  static uuid1() {
    return uuid.v1();
  }
}

Neo4js.prototype.Error = Neo4js.Error = neo4jsErrors.BaseError;
Neo4js.prototype.Model = Neo4js.Model = Model;
Neo4js.prototype.ModelObject = Neo4js.ModelObject = ModelObject;

export default Neo4js;
