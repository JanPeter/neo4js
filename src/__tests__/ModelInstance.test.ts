// @flow

import neo4js, {
  Model,
  ModelInstance,
  hasMany,
  model,
  relation,
  hasOne,
  defaultProps,
  StringProperty,
  NumberProperty,
  HasManyActions,
  HasOneActions,
} from "../index";

type PersonProps = {
  name: StringProperty;
};

type TaskProps = {
  title: StringProperty;
  done?: boolean;
};

type TodoProps = TaskProps;

const PersonTaskRelation = relation
  .from(() => Person)
  .to(() => Task)
  .via("created");

const FriendRelation = relation
  .from(() => Person)
  .to(() => Person)
  .via("friend");

class PersonModel extends Model<PersonProps, PersonInstance> {}
const Person: PersonModel = new PersonModel("Person");

class TaskModel extends Model<TaskProps, TaskInstance> {}
const Task: TaskModel = new TaskModel("Task");

@model(Person)
class PersonInstance extends ModelInstance<PersonProps> {
  @hasMany(Task, PersonTaskRelation)
  tasks: HasManyActions<TaskProps, TaskInstance>;
  @hasMany(Person, FriendRelation, "any")
  friends: HasManyActions<PersonProps, PersonInstance>;

  getFriendsWithName = async (name: string) => {
    return await this.friends.get({ name: { $sw: name } });
  };
}

@defaultProps({
  title: "(empty)",
  done: () => true,
})
@model(Task)
class TaskInstance extends ModelInstance<TaskProps> {}

const Todo = new Model("Todo");
@defaultProps({
  title: "(empty)",
  done: () => false,
})
@model(Todo)
class TodoInstance extends ModelInstance<TodoProps> {}

describe("ModelInstance", () => {
  beforeAll(() => {
    neo4js.init({
      boltUri: "localhost",
      boltPort: 10001,
    });
  });

  beforeEach(async () => {
    const paul = await Person.create({ name: "Olaf" });
    await paul.friends.create([{ name: "Olga" }, { name: "Hannes" }], {
      since: "yesterday",
      location: "Graz",
    });
    await Task.create({ title: "Learn magic", done: true });
    await Task.create({ title: "Write more test cases", done: false });
    await Task.create({ title: "Write more test cases", done: false });
  });

  afterEach(async () => {
    await neo4js.run("MATCH (n) DETACH DELETE n");
  });

  afterAll(() => {
    neo4js.close();
  });

  it("should return an instance of PersonInstance", async () => {
    const paul: PersonInstance = await Person.create({ name: "Paul" });

    expect(paul).toBeInstanceOf(PersonInstance);
  });

  describe("ModelInstance functions", () => {
    it("should return friends with starting name 'O'", async () => {
      const olaf: PersonInstance = await Person.findOne({ name: "Olaf" });
      if (!olaf) {
        expect(1).toEqual(0);
        return;
      }
      let friends = await olaf.getFriendsWithName("O");
      friends = friends.map(f => {
        delete f.props.guid;
        return f;
      });
      expect(friends).toMatchSnapshot();
    });
  });

  describe("Default Properties", () => {
    it("should initialize with default properties", async () => {
      let task: TaskInstance = await Task.create({
        title: "Write more tests boi!",
      });
      const guid = task.props.guid;

      if (task && task.props.guid) delete task.props.guid;
      expect(task).toMatchSnapshot();

      task = await Task.findByGuid(guid);
      if (task && task.props.guid) delete task.props.guid;
      expect(task).toMatchSnapshot();
    });

    it("should initialize with minimal configuration default properties", async () => {
      let todo = await Todo.create({});
      delete todo.props.guid;
      expect(todo).toMatchSnapshot();
    });
  });
});
