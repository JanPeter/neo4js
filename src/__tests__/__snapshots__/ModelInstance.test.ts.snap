// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelInstance Default Properties should initialize with default properties 1`] = `
TaskInstance {
  "props": Object {
    "done": true,
    "title": "Write more tests boi!",
  },
}
`;

exports[`ModelInstance Default Properties should initialize with default properties 2`] = `
TaskInstance {
  "props": Object {
    "done": true,
    "title": "Write more tests boi!",
  },
}
`;

exports[`ModelInstance Default Properties should initialize with minimal configuration default properties 1`] = `
TodoInstance {
  "props": Object {
    "done": false,
    "title": "(empty)",
  },
}
`;

exports[`ModelInstance ModelInstance functions should return friends with starting name 'O' 1`] = `
Array [
  PersonInstance {
    "friends": Object {
      "add": [Function],
      "count": [Function],
      "create": [Function],
      "get": [Function],
      "remove": [Function],
      "update": [Function],
    },
    "getFriendsWithName": [Function],
    "props": Object {
      "name": "Olga",
    },
    "relationProps": Object {
      "location": "Graz",
      "since": "yesterday",
    },
    "tasks": Object {
      "add": [Function],
      "count": [Function],
      "create": [Function],
      "get": [Function],
      "remove": [Function],
      "update": [Function],
    },
  },
]
`;
